### PR TITLE
Date fixes for `Birthday`, `Meeting` and `SpecialDate`

### LINKED ISSUE #178: [PE-D] Able to add invalid birthdate
Able to add a birthdate of a person past the current date

-------------
Labels: `severity.Medium` `type.FeatureFlaw`
original: branzuelajohn/ped#2

### LINKED ISSUE #189: [PE-D] Able to edit birthdate past current date
No details provided.

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: branzuelajohn/ped#13

### LINKED ISSUE #191: [PE-D] Able to add invalid meeting dates
Eg. birthday of person is in 1998 but able to set meeting on 1997

-------------
Labels: `severity.Low` `type.FunctionalityBug`
original: branzuelajohn/ped#15

### PR DESCRIPTION
**Pull request by ivantjh**
- A `birthday` is only valid if it is before or on today
- A `specialDate` is only valid if it is before or on today and that it happens after the person is born
- A `meeting` is only valid if the meeting occurs before the current time or now and that it happens after the person is born (i.e. it does checks based on time too)
- Refactored `Event` to have `Meeting` and `SpecialDate` subclass from it (this makes the json validation stricter because we won't have situations where we have a specialDate `Event` but there is a `time` in it)
- Add in tests to guard against boundary values
- Most of the code is from the subclassing

Birthday: Fixes #178, fixes #189
Meeting: Fixes #191

### CODE REVIEW

File: src/main/java/seedu/address/model/person/Birthday.java [hunk 1]
```diff
@@ -37,35 +37,42 @@ public Birthday(String birthday) {
...
 
     public LocalDate getDate() {
-        return date;
+        return birthDate;
     }
 
     /**
      * Returns true if a given string is a valid birthday.
      */
     public static boolean isValidBirthday(String test) {
```
Comment (chesterhow) [inline src/main/java/seedu/address/model/person/Birthday.java]: Can we rename the param `test`?

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 2]
```diff
@@ -107,7 +110,7 @@ public void parse_compulsoryFieldMissing_failure() {
         assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + VALID_ADDRESS_BOB
                 + BIRTHDAY_DESC_BOB, expectedMessage);
 
-        // missing address prefix
+        // missing birthday prefix
```
Comment (chesterhow) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:113]: Just realised this test and the previous test is exactly the same.

File: src/main/java/seedu/address/logic/commands/AddMeetingCommand.java [hunk 3]
```diff
@@ -35,30 +37,31 @@
             + PREFIX_TIME + "1240 "
             + PREFIX_DESCRIPTION + "We went to the beach!";
 
+    public static final String MESSAGE_CONSTRAINTS =
+            "Meetings should have already happened before they are recorded. Meeting description should not be empty";
```
Comment (chesterhow) [inline src/main/java/seedu/address/logic/commands/AddMeetingCommand.java]: Can reword to "Meetings are records of past events, and should not be in the future. Meeting description should not be empty."

**Review by chesterhow — APPROVED** [review]
LGTM

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java (hunk 2 above)
Comment (ivantjh) [reply src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:113]: LOL, i just realised it too.
